Add background and audio checks

Naxesss opened the pull request:
Makes more progress towards resolving https://github.com/ppy/osu/issues/12091

This adds the following:
- `CheckBackgroundQuality`, which checks
    - 160x120 <= resolution <= 2560x1440
    - filesize <= 2.5 MB
- `CheckAudioQuality`, which checks
    - 128 <= average bitrate <= 192
- `CheckAudioPresence` (audio equivalent of `CheckBackgroundPresence`, both inheriting `CheckFilePresence`)
- Tests for all the new checks
- An `IWorkingBeatmap` argument to `Run` of `ICheck`, which enables retrieving of
    - the background texture (`Background`)
    - the audio track (`Track`)
- `Track` and `GetStream` to `IWorkingBeatmap`
- `GetPathForFile` to `BeatmapSetInfo`

Naxesss committed:
Use `WorkingBeatmap` instead of `IBeatmap`

This lets us access things like the background, track, etc. which are necessary for quality and filesize checks.

Also improves the structure of the `CheckBackgroundTest` class in the process.

Naxesss committed:
Add `GetStream` to `IWorkingBeatmap`

This is necessary to obtain the filesize of the audio and background files.

Naxesss committed:
Add `GetPathForFile` to `BeatmapSetInfo`

This is used in several places, and so should probably have a function rather than remaining as duplicated code.

Also applies this together with the previous commit to `BeatmapManagerWorkingBeatmap`.

Naxesss committed:
Add check for background resolution and filesize

Naxesss committed:
Don't return low res and too low res at the same time

Naxesss committed:
Add `GetStream` to `IWorkingBeatmap`

Naxesss committed:
Fix resolved fields

Naxesss committed:
Clean up check logic

Makes use of the new `BeatmapSet.GetPathForFile` method and removes dependency on `WorkingBeatmap` specifically, allowing us to switch to `IWorkingBeatmap` later.

Naxesss committed:
Take `IWorkingBeatmap` instead of `WorkingBeatmap`

This makes testing much easier, and allows for checking of any class deriving from that interface, including `WorkingBeatmap`.

Naxesss committed:
Add background quality check tests

Naxesss committed:
Factor out general file presence checking

This allows us to use the same method of checking for other files that should exist, for example the audio file.

By using the same method, they all share test cases too.

Naxesss committed:
Add audio presence check

Naxesss committed:
Add the playable beatmap as check argument

This is different from the working beatmap's `.Beatmap` property in that it is mutated by the ruleset/editor.

So hit objects, for example, are actually of type `Slider` and such instead of the legacy `ConvertSlider`.

This should be preferred over `workingBeatmap.Beatmap`.

Naxesss committed:
Simplify background quality test names

Naxesss committed:
Use the playable beatmap for file presence checks

Naxesss committed:
Show 2 decimals for background filesize

Naxesss committed:
Add `Track` as a property to `IWorkingBeatmap`

This is implemented by `WorkingBeatmap` already, and is much better to use than loading the track every time we need it.

Naxesss committed:
Add audio quality check

Naxesss committed:
Add audio quality check tests

Naxesss committed:
Fix category of audio quality check

Naxesss committed:
Add xmldoc to `GetStream`

Naxesss committed:
Improve xmldoc consistency

Naxesss committed:
Merge branch 'master' into bg-audio-quality-checks

Naxesss committed:
Remove now unused import

smoogipoo commented on osu.Game.Tests/Editing/Checks/CheckAudioQualityTest.cs:
```diff
@@ -0,0 +1,114 @@
+            beatmap.Metadata.AudioFile = null;
+
+            var mock = new Mock<IWorkingBeatmap>();
+            mock.SetupGet(_ => _.Beatmap).Returns(beatmap);
```
In this project, `_` imply discards, not usages. Please name these (like `b` or `w` for workingbeatmap, `t` for track, etc).
This is done in multiple places in this PR, so please change all.

smoogipoo commented on osu.Game/Screens/Edit/Verify/VerifyScreen.cs:
```diff
@@ -118,10 +122,11 @@ protected override void LoadComplete()
 
                 if (rulesetVerifier != null)
-                    issues = issues.Concat(rulesetVerifier.Run(Beatmap));
+                    issues = issues.Concat(rulesetVerifier.Run(beatmap.PlayableBeatmap, workingBeatmap));
```
Any reason for passing `EditorBeatmap.PlayableBeatmap` through instead of `EditorBeatmap`? Internally it forwards everything, but it's probably better this way to keep things consistent?

smoogipoo reviewed (changes requested):
Rest looks solid

peppy commented on osu.Game/Beatmaps/BeatmapSetInfo.cs:
```diff
@@ -59,6 +59,8 @@ public class BeatmapSetInfo : IHasPrimaryKey, IHasFiles<BeatmapSetFileInfo>, ISo
 
         public string StoryboardFile => Files?.Find(f => f.Filename.EndsWith(".osb", StringComparison.OrdinalIgnoreCase))?.Filename;
 
+        public string GetPathForFile(string filename) => Files?.SingleOrDefault(f => string.Equals(f.Filename, filename, StringComparison.OrdinalIgnoreCase))?.FileInfo.StoragePath;
```
Best to add xmldoc for this (generally we try and xmldoc anything that is `public`) with a mention that the returned path is relative to the user file storage.

peppy commented on osu.Game/Screens/Edit/Verify/VerifyScreen.cs:
```diff
@@ -118,10 +122,11 @@ protected override void LoadComplete()
             private void refresh()
             {
-                var issues = generalVerifier.Run(Beatmap);
+                var workingBeatmap = beatmapManager.GetWorkingBeatmap(beatmap.BeatmapInfo);
```
You can get the `WorkingBeatmap` via DI rather than re-requesting:

```csharp
        [Resolved]
        private IBindable<WorkingBeatmap> working { get; set; }
```

smoogipoo replied:
As a side note, I thought about this too, because the working beatmap instance isn't requeried by the editor when it's changed. The working beatmap cache is invalidated (see: `BeatmapManager.Save()`), but that's all.

There's a potential for discrepancies if, e.g. `BeatmapInfo` is updated on the wrong instance, because:
1. `Editor.Save()` uses the playable beatmap to save in all cases.
2. `PlayableBeatmap.BeatmapInfo` and `WorkingBeatmap.BeatmapInfo` are different instances (see: `WorkingBeatmap.GetPlayableBeatmap()`).

Just something to take note of and probably fix up - I don't like that the sections in `SetupScreen` are very careful about which beatmap to use because of this. I don't think reuse of the working beatmap causes any problems in this case...

peppy replied:
Yeah, definitely.

There's a hope that when switching to realm and refactoring this logic, these will actually be referencing the same instance.

smoogipoo replied:
I'm not sure about that - it's done this way specifically to support mods that are `IApplicableToDifficulty`, so unless that was refactored out `BeatmapInfo`/`IBeatmap`...

Naxesss replied:
Fixed, and also double checked that editing the map and its resources properly updated all checks.

Because I use the playable beatmap in all cases apart from retrieving background texture / audio track, there shouldn't be an issues with getting an outdated `BeatmapInfo`, so reusing is probably fine. Will keep that in mind, though.

peppy reviewed (changes requested):
Solid apart from a few minor issues.

Naxesss committed:
Fix wrong assert in offscreen test

Naxesss committed:
Add file hash to file presence test

Necessary because we now find the storage path of the file rather than just the file itself.

Naxesss committed:
Fix usage of _ in Moq lambdas

Naxesss committed:
Simplify verifier run call args

Uses the resolved working beatmap instead of resolving it every time.

Also uses the EditorBeatmap itself as playable beatmap, as it is of type `IBeatmap` already, and `.PlayableBeatmap` forwards everything anyway.

Naxesss committed:
Add xmldoc to `GetPathForFile`

Naxesss committed:
Add xmldoc note about path being relative
